en/_posts/2024-05-30-understanding-github-artifact-attestations.md

```diff
@@ -0,0 +1,311 @@
+          github-hosted   .
+       1.3.6.1.4.1.57264.1.12:
+          .:https://github.com/ianlewis/gha-artifact-attestations-test
+   ...
+   ```
+
+3. A SLSA
+   [predicate](https://github.com/in-toto/attestation/blob/main/spec/v1/predicate.md)
+   is generated and the provenance statement is signed with the returned
+   certificate’s private key. The resulting signature is combined with the
+   provenance to create a full attestation bundle and this bundle is recorded
+   in GitHub’s attestation store.
+
+   The SLSA predicate looks something like this:
+
+   ```json
+   {
+     "buildDefinition": {
+       "buildType": "https://slsa-framework.github.io/github-actions-buildtypes/workflow/v1",
+       "externalParameters": {
+         "workflow": {
+           "ref": "refs/heads/main",
+           "repository": "https://github.com/ianlewis/gha-artifact-attestations-test",
+           "path": ".github/workflows/artifact-attestations.basic.yml"
+         }
+       },
+       "internalParameters": {
+         "github": {
+           "event_name": "push",
+           "repository_id": "803607921",
+           "repository_owner_id": "49289"
+         }
+       },
+       "resolvedDependencies": [
+         {
+           "uri": "git+https://github.com/ianlewis/gha-artifact-attestations-test@refs/heads/main",
+           "digest": {
+             "gitCommit": "9e68bb76632788dc01c6596298fb015f46b7fe0f"
+           }
+         }
+       ]
+     },
+     "runDetails": {
+       "builder": {
+         "id": "https://github.com/actions/runner/github-hosted"
+       },
+       "metadata": {
+         "invocationId": "https://github.com/ianlewis/gha-artifact-attestations-test/actions/runs/9171197474/attempts/1"
+       }
+     }
+   }
+   ```
+
+I'm leaving out some details but this is the general flow for attestation. So as
+a result we have _both_ an attestation bundle in JSON format \_and \_this
+includes the Sigstore certificate with some OID claims set.
+
+After we have an artifact and attestation, as a user, we need to be able to
+verify it before we use it. Verification works something like the following. The
+code for this is found in the [`gh attestations verify`
+command](https://github.com/cli/cli/tree/trunk/pkg/cmd/attestation/verify)
+for GitHub’s official CLI tool.
+
+1. The attestation is downloaded from the GitHub Attestations API using the
+   artifact’s digest.
+2. The attestation’s signature is verified against the public key used to sign
+   it, and to the certificate chain for either the GitHub Fulcio instance, or
+   the public Sigstore Fulcio instance (the root certificates are managed using
+   [TUF](https://theupdateframework.io/)).
+3. The expected values for the owner or repo given by the user are matched
+   against the signing certificate’s OID claims.
+
+Notice that nowhere here did we actually use the contents of the SLSA predicate
+for verification. I think this is an oversight but we’ll discuss why that might
```

Comment:
Add a note that the attestation store performs these checks.

Reply:
I'm going to leave discussion about these checks out and produce a separate PR for the `gh` CLI.